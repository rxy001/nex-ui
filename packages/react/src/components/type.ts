export type ClassNames = string | string[]

export type ComponentClasses<OwnerState, T extends string> = Partial<
  Record<T, ClassNames | ((ownerState: OwnerState) => ClassNames)>
>
