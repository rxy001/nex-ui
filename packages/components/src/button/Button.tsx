import { WaterWave } from '../waterWave'
import type { ButtonProps } from './types'
import { useButton } from './useButton'

export function Button(props: ButtonProps) {
  const { children, disabled, type, href } = props

  const { getProps } = useButton(props)

  const rootProps = {
    ...getProps(),
  }

  if (href !== undefined) {
    return (
      <a href={href} data-disabled={disabled} {...rootProps}>
        {children}
      </a>
    )
  }

  return (
    <WaterWave>
      <button type={type} disabled={disabled} {...rootProps}>
        {children}
      </button>
    </WaterWave>
  )
}
