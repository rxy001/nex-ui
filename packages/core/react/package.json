{
  "name": "@ant-ui/react",
  "version": "1.0.0",
  "author": "X1ng Yu",
  "scripts": {},
  "main": "src/index.ts",
  "dependencies": {
  }
}
