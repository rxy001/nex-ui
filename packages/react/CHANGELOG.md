# @nex-ui/react

## 0.0.15

### Patch Changes

- fdd4483: Release
- Updated dependencies [fdd4483]
  - @nex-ui/styled@0.0.15
  - @nex-ui/system@0.0.15
  - @nex-ui/hooks@0.0.15
  - @nex-ui/icons@0.0.15
  - @nex-ui/utils@0.0.15

## 0.0.14

### Patch Changes

- ca25983: Release
- Updated dependencies [ca25983]
  - @nex-ui/styled@0.0.14
  - @nex-ui/system@0.0.14
  - @nex-ui/hooks@0.0.14
  - @nex-ui/icons@0.0.14
  - @nex-ui/utils@0.0.14

## 0.0.13

### Patch Changes

- 1b97788: Release
- Updated dependencies [1b97788]
  - @nex-ui/styled@0.0.13
  - @nex-ui/system@0.0.13
  - @nex-ui/hooks@0.0.13
  - @nex-ui/icons@0.0.13
  - @nex-ui/utils@0.0.13

## 0.0.12

### Patch Changes

- 664eb21: Test

## 0.0.11

### Patch Changes

- 259356e: Test
- Updated dependencies [259356e]
  - @nex-ui/styled@0.0.11
  - @nex-ui/system@0.0.11
  - @nex-ui/hooks@0.0.11
  - @nex-ui/icons@0.0.11
  - @nex-ui/utils@0.0.10

## 0.0.10

### Patch Changes

- 5a229bd: release
- Updated dependencies [5a229bd]
  - @nex-ui/styled@0.0.10
  - @nex-ui/system@0.0.10
  - @nex-ui/icons@0.0.10

## 0.0.9

### Patch Changes

- 4a5c359: release
- Updated dependencies [4a5c359]
  - @nex-ui/styled@0.0.9
  - @nex-ui/system@0.0.9
  - @nex-ui/hooks@0.0.9
  - @nex-ui/icons@0.0.9
  - @nex-ui/utils@0.0.9

## 0.0.8

### Patch Changes

- c8c1956: release
- Updated dependencies [c8c1956]
  - @nex-ui/styled@0.0.8
  - @nex-ui/system@0.0.8
  - @nex-ui/hooks@0.0.8
  - @nex-ui/icons@0.0.8
  - @nex-ui/utils@0.0.8

## 0.0.7

### Patch Changes

- 2618e1a: release
- Updated dependencies [2618e1a]
  - @nex-ui/styled@0.0.7
  - @nex-ui/system@0.0.7
  - @nex-ui/hooks@0.0.7
  - @nex-ui/icons@0.0.7
  - @nex-ui/utils@0.0.7

## 0.0.6

### Patch Changes

- 87191c1: release
- Updated dependencies [87191c1]
  - @nex-ui/styled@0.0.6
  - @nex-ui/system@0.0.6
  - @nex-ui/hooks@0.0.6
  - @nex-ui/icons@0.0.6
  - @nex-ui/utils@0.0.6

## 0.0.5

### Patch Changes

- 7631577: release
- Updated dependencies [7631577]
  - @nex-ui/styled@0.0.5
  - @nex-ui/system@0.0.5
  - @nex-ui/hooks@0.0.5
  - @nex-ui/icons@0.0.5
  - @nex-ui/utils@0.0.5

## 0.0.4

### Patch Changes

- 70cb478: v0.0.4
- Updated dependencies [70cb478]
  - @nex-ui/styled@0.0.4
  - @nex-ui/system@0.0.4
  - @nex-ui/hooks@0.0.4
  - @nex-ui/icons@0.0.4
  - @nex-ui/utils@0.0.4
