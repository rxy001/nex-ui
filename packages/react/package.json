{
  "name": "@nex-ui/react",
  "version": "0.0.12",
  "author": "X1ng Yu",
  "description": "ðŸŽ‰ A beautiful, modern, and reliable React component library.",
  "homepage": "https://nex-ui-docs.vercel.app/",
  "private": false,
  "keywords": [
    "react",
    "react-component",
    "nui",
    "nex-ui"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rxy001/nex-ui.git",
    "directory": "packages/react"
  },
  "license": "MIT",
  "scripts": {
    "build": "tsx ../../scripts/build.ts --dts",
    "build:fast": "tsx ../../scripts/build.ts",
    "typecheck": "tsc --noEmit",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "type": "module",
  "files": [
    "dist"
  ],
  "main": "src/index.ts",
  "devDependencies": {
    "@storybook/react": "8.6.12"
  },
  "dependencies": {
    "@nex-ui/hooks": "workspace:*",
    "@nex-ui/icons": "workspace:*",
    "@nex-ui/styled": "workspace:*",
    "@nex-ui/system": "workspace:*",
    "@nex-ui/utils": "workspace:*",
    "clsx": "2.1.1"
  },
  "peerDependencies": {
    "@emotion/react": ">=11",
    "motion": "^12.7.2",
    "react": ">=19",
    "react-dom": ">=19"
  },
  "sideEffects": false,
  "clean-package": "../../clean-package.config.json"
}
