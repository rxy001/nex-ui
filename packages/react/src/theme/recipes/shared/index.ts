export * from './colorVariant'
export * from './radiusVariant'
export * from './toSlot'
export * from './fullWidthVariant'
