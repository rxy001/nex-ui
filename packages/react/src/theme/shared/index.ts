export * from './colorVariant'
export * from './radiusVariant'
export * from './toSlots'
export * from './fullWidthVariant'
export * from './sizeVariant'
