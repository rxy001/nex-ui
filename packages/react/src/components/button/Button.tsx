import { forwardRef } from 'react'
import type { Ref } from 'react'
import { WaterWave } from '../waterWave'
import type { ButtonProps } from './types'
import { useButton } from './useButton'

export const Button = forwardRef<
  HTMLButtonElement | HTMLAnchorElement,
  ButtonProps
>((props, ref) => {
  const { loading } = props

  const { getRootProps, startIcon, endIcon } = useButton(props)

  const { type, href, disabled, children, ...rootProps } = getRootProps()

  const childNode =
    href !== undefined ? (
      <a
        ref={ref as Ref<HTMLAnchorElement>}
        data-disabled={disabled || null}
        href={href}
        {...rootProps}
      >
        {startIcon}
        {children}
        {endIcon}
      </a>
    ) : (
      <button
        type={type}
        ref={ref as Ref<HTMLButtonElement>}
        disabled={disabled}
        {...rootProps}
      >
        {startIcon}
        {children}
        {endIcon}
      </button>
    )

  return <WaterWave disabled={loading || disabled}>{childNode}</WaterWave>
})
