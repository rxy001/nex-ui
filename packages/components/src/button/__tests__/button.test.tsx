import { describe, it, expect, jest } from '@jest/globals'
import { mountTest, refTest } from '@/tests/shared'
import { useState } from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Button } from '../Button'
import type { ButtonProps } from '../types'

describe('Button', () => {
  mountTest(Button)
  refTest(Button)

  it('renders correctly', () => {
    const { container } = render(<Button>Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should trigger onClick function', () => {
    const onClick = jest.fn()
    const { container } = render(<Button onClick={onClick} />)
    fireEvent.click(container.firstChild!)
    expect(onClick).toHaveBeenCalled()
  })

  it('should ignore events when disabled', () => {
    const onClick = jest.fn()
    const { getByText } = render(
      <>
        <Button disabled onClick={onClick}>
          Btn Tag
        </Button>
        <Button disabled onClick={onClick} href="#">
          A Tag
        </Button>
      </>,
    )
    fireEvent.click(getByText('Btn Tag'))
    fireEvent.click(getByText('A Tag'))
    expect(onClick).not.toHaveBeenCalled()
  })

  it('should support link button', () => {
    const { container } = render(
      <Button target="_blank" href="https://">
        Link Button
      </Button>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should renders with start icon', () => {
    const { getByTestId } = render(
      <Button startIcon={<span data-testid="start-icon">Icon</span>}>
        Button
      </Button>,
    )
    expect(getByTestId('start-icon')).toBeInTheDocument()
  })

  it('should renders with end icon', () => {
    const { getByTestId } = render(
      <Button startIcon={<span data-testid="end-icon">Icon</span>}>
        Button
      </Button>,
    )
    expect(getByTestId('end-icon')).toBeInTheDocument()
  })

  it('should support to change loading', () => {
    const DefaultButton: React.FC = () => {
      const [loading, setLoading] = useState<ButtonProps['loading']>(false)
      return (
        <Button loading={loading} onClick={() => setLoading(true)}>
          Button
        </Button>
      )
    }
    const { container } = render(<DefaultButton />)
    fireEvent.click(container.firstChild!)
    expect(container.querySelector('.nexui-start-icon')).toBeInTheDocument()
  })
})
