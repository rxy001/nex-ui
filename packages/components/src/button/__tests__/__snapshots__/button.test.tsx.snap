// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  class="nexui-btn"
  type="button"
>
  Button
</button>
`;

exports[`Button should support link button 1`] = `
<a
  class="nexui-btn"
  href="https://"
  target="_blank"
>
  Link Button
</a>
`;
